src/libraries/Microsoft.PowerFx.Core/Public/Values/CollectionTableValue.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using Microsoft.PowerFx.Core.IR;
+
+namespace Microsoft.PowerFx.Core.Public.Values
+{
+    /// <summary>
+    /// Create a TableValue over a dotnet collection class.
+    /// Depending on what collection interfaces it exposes will determine the capabilities 
+    /// (such as Enumeration, Count, Index).
+    /// </summary>
+    /// <typeparam name="T">The element type of the collection.</typeparam>
+    internal abstract class CollectionTableValue<T> : TableValue
+    {
+        // Required - basic enumeration. This is the source object. 
+        private readonly IEnumerable<T> _enumerator; // required. supports enumeration;
+
+        // Additional capabilities. 
+        private readonly IReadOnlyList<T> _sourceIndex; // maybe null. supports index. 
+        private readonly IReadOnlyCollection<T> _sourceCount; // maybe null. supports count;
+
+        /// <summary>
+        /// This table can be enumerated. 
+        /// </summary>
+        public bool CanEnumerate => _enumerator != null;
+
+        /// <summary>
+        /// This table can be directly indexed (without a linear scan).
+        /// </summary>
+        public bool CanIndex => _sourceIndex != null;
+
+        /// <summary>
+        /// This table can return a count (without a linear scan). 
+        /// </summary>
+        public bool CanCount => _sourceCount != null;
```

Comment:
>CanCount [](http://example.com/codeflow?start=20&length=8)

Typically if types have a CanXXX member and a XXX member, if CanXXX returns false then XXX will fail / throw (example: Stream.CanRead/CanWrite and Read/Write), which is not the case here...

Reply:
Are these properties (CanIndex / CanCount) needed in the first place?

Reply:
We could remove.  It's telling if the ability is O[1] or O[N]. 
This is similar to delegation capabilities. You "can" do an operation that 's not delegated, it's just linear & on client.

Reply:
I'd rather remove them to reduce the surface area, and add them in the future if we need them